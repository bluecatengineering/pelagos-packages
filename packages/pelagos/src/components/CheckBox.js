import {useEffect, useRef} from 'react';
import PropTypes from 'prop-types';

import './CheckBox.less';

/* A checkbox. */
const CheckBox = ({className, label, indeterminate, disabled, error, ...props}) => {
	const inputRef = useRef();
	useEffect(() => (inputRef.current.indeterminate = indeterminate), [indeterminate]);
	return (
		<label className={`CheckBox${disabled ? ' CheckBox--disabled' : ''}${className ? ` ${className}` : ''}`}>
			<input
				{...props}
				className={error ? 'error' : ''}
				type="checkbox"
				disabled={disabled}
				aria-invalid={error}
				ref={inputRef}
			/>
			{label}
		</label>
	);
};

CheckBox.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The checkbox label. */
	label: PropTypes.node,
	/** Whether the box is checked. */
	checked: PropTypes.bool,
	/** Whether the box state is indeterminate. */
	indeterminate: PropTypes.bool,
	/** Whether the box is disabled. */
	disabled: PropTypes.bool,
	/** Whether the box is in error. */
	error: PropTypes.bool,
	/** Function invoked when the checked status is changed. */
	onChange: PropTypes.func,
};

export default CheckBox;
