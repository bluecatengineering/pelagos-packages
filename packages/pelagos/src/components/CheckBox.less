@import '../../less/spacing';
@import '../../less/utils';

.CheckBox {
	@layer-hover-effects();
	// stylelint-disable-next-line @bluecateng/property-strict-value -- minimum space to fit focus ring
	margin-left: 3px;

	&--disabled {
		color: var(--text-disabled);

		&:hover,
		&:active {
			background-color: transparent;
		}
	}

	& > input {
		margin-right: @sp-08;
	}
}
