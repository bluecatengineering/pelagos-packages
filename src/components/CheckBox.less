@import '../../less/focus';
@import '../../less/spacing';
@import '../../less/utils';

.CheckBox {
	@layer-hover-effects();
	display: inline-block;
	position: relative;
	padding-left: @sp-24;
	margin-left: @sp-02;
	cursor: pointer;
	outline: none;
}

.CheckBox::before,
.CheckBox::after {
	display: block;
	position: absolute;
	content: '';
}

.CheckBox::before {
	top: @sp-02;
	left: 0;
	width: 16px;
	height: 16px;
	border: 1px solid var(--icon-primary);
	border-radius: 2px;

	@media print {
		border-color: var(--print-color);
	}
}
.CheckBox[aria-checked='true']::before,
.CheckBox[aria-checked='mixed']::before {
	background-color: var(--interactive);
	border-color: var(--interactive);

	@media print {
		background-color: transparent;
		border-color: var(--print-medium);
	}
}

.CheckBox--error::before {
	border-color: var(--support-error);
}
.CheckBox--error[aria-checked='true']::before,
.CheckBox--error[aria-checked='mixed']::before {
	background-color: var(--support-error);
	border-color: var(--support-error);
}

.CheckBox:focus::before {
	@focus-ring();
	outline-offset: 1px;
}
.CheckBox--error:focus::before {
	@focus-ring-error();
	outline-offset: 1px;
}

.CheckBox[aria-checked='true']::after {
	top: 3px;
	left: 5px;
	width: 7px;
	height: 12px;
	border: solid var(--text-inverse);
	border-width: 0 3px 3px 0;
	transform: rotate(45deg);

	@media print {
		border-color: var(--print-color);
	}
}

.CheckBox[aria-checked='mixed']::after {
	top: 8px;
	left: 3px;
	width: 10px;
	border-bottom: 3px solid var(--text-inverse);

	@media print {
		border-color: var(--print-color);
	}
}

.CheckBox[aria-disabled='true'] {
	color: var(--text-disabled);
	cursor: not-allowed;
	pointer-events: none;
}
.CheckBox[aria-disabled='true']::before {
	border-color: var(--icon-disabled);
}
.CheckBox[aria-disabled='true'][aria-checked='true']::before,
.CheckBox[aria-disabled='true'][aria-checked='mixed']::before {
	background-color: var(--icon-disabled);
}
