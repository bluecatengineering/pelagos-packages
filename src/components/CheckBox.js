import {useEffect, useRef} from 'react';
import PropTypes from 'prop-types';

import './CheckBox.less';

/* A checkbox. */
const CheckBox = ({className, label, indeterminate, error, ...props}) => {
	const inputRef = useRef();
	useEffect(() => (inputRef.current.indeterminate = indeterminate), [indeterminate]);
	return (
		<label className={`CheckBox${className ? ` ${className}` : ''}`}>
			<input {...props} className={error ? 'error' : ''} type="checkbox" ref={inputRef} />
			{label}
		</label>
	);
};

CheckBox.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The check box label. */
	label: PropTypes.node,
	/** Whether the box is checked. */
	checked: PropTypes.bool,
	/** Whether the box state is indeterminate. */
	indeterminate: PropTypes.bool,
	/** Whether the box is disabled. */
	disabled: PropTypes.bool,
	/** Whether the box is in error. */
	error: PropTypes.bool,
	/** Function invoked when the checked status is changed. */
	onChange: PropTypes.func,
};

export default CheckBox;
