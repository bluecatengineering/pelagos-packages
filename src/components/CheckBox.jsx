import React, {useCallback} from 'react';
import PropTypes from 'prop-types';

import './CheckBox.less';

const CheckBox = ({id, componentId, className, label, checked, disabled, error, onChange}) => {
	const handleClick = useCallback(event => (disabled ? null : onChange(event)), [disabled, onChange]);

	const handleKeyDown = useCallback(
		event => (disabled || event.keyCode !== 32 ? null : (event.preventDefault(), onChange(event))),
		[disabled, onChange]
	);

	return (
		<span
			id={id}
			data-bcn-id={componentId}
			className={'CheckBox ' + className + (!disabled && error ? ' CheckBox--error' : '')}
			role="checkbox"
			aria-disabled={disabled}
			aria-checked={!!checked}
			tabIndex={disabled ? -1 : 0}
			onClick={handleClick}
			onKeyDown={handleKeyDown}>
			{label}
		</span>
	);
};

CheckBox.propTypes = {
	id: PropTypes.string,
	componentId: PropTypes.string,
	className: PropTypes.string,
	label: PropTypes.string,
	checked: PropTypes.bool,
	disabled: PropTypes.bool,
	error: PropTypes.bool,
	onChange: PropTypes.func,
};

export default CheckBox;
