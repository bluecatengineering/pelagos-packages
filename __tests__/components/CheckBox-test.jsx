import React from 'react';
import {shallow} from 'enzyme';

import CheckBox from '../../src/components/CheckBox';

jest.unmock('../../src/components/CheckBox');

describe('CheckBox', () => {
	describe('rendering', () => {
		it('renders expected elements', () => {
			const wrapper = shallow(
				<CheckBox id="test" componentId="test" className="TestClass" label="Test" checked onChange={jest.fn()} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when disabled is true', () => {
			const wrapper = shallow(
				<CheckBox componentId="test" className="TestClass" label="Test" checked disabled onChange={jest.fn()} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when error is true', () => {
			const wrapper = shallow(
				<CheckBox componentId="test" className="TestClass" label="Test" checked error onChange={jest.fn()} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onChange when clicked', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<CheckBox className="TestClass" label="Test" checked onChange={onChange} />);
			wrapper.simulate('click');
			expect(onChange).toHaveBeenCalled();
		});

		it('does not call onChange when clicked if disabled is true', () => {
			const onChange = jest.fn();
			const wrapper = shallow(<CheckBox className="TestClass" label="Test" checked disabled onChange={onChange} />);
			wrapper.simulate('click');
			expect(onChange).not.toHaveBeenCalled();
		});

		it('calls onChange when the space key is pressed', () => {
			const onChange = jest.fn();
			const event = {preventDefault: jest.fn(), keyCode: 32};
			const wrapper = shallow(<CheckBox className="TestClass" label="Test" checked onChange={onChange} />);
			wrapper.simulate('keydown', event);
			expect(onChange).toHaveBeenCalled();
			expect(event.preventDefault).toHaveBeenCalled();
		});

		it('does not calls onChange when the space key is pressed if disabled is true', () => {
			const onChange = jest.fn();
			const event = {preventDefault: jest.fn(), keyCode: 32};
			const wrapper = shallow(<CheckBox className="TestClass" label="Test" checked disabled onChange={onChange} />);
			wrapper.simulate('keydown', event);
			expect(onChange).not.toHaveBeenCalled();
			expect(event.preventDefault).not.toHaveBeenCalled();
		});

		it('does not call onChange when the any other key is pressed', () => {
			const onChange = jest.fn();
			const event = {preventDefault: jest.fn(), keyCode: 9};
			const wrapper = shallow(<CheckBox className="TestClass" label="Test" checked onChange={onChange} />);
			wrapper.simulate('keydown', event);
			expect(onChange).not.toHaveBeenCalled();
			expect(event.preventDefault).not.toHaveBeenCalled();
		});
	});
});
