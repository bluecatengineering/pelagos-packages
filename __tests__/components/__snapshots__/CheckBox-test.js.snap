// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox rendering renders expected elements 1`] = `
<span
  aria-checked={true}
  className="CheckBox TestClass"
  id="test"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex={0}
>
  Test
</span>
`;

exports[`CheckBox rendering renders expected elements when disabled is true 1`] = `
<span
  aria-checked={true}
  aria-disabled={true}
  className="CheckBox TestClass"
  id="test"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex={-1}
>
  Test
</span>
`;

exports[`CheckBox rendering renders expected elements when error is true 1`] = `
<span
  aria-checked={true}
  className="CheckBox TestClass CheckBox--error"
  id="test"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex={0}
>
  Test
</span>
`;

exports[`CheckBox rendering renders expected elements when indeterminate is true 1`] = `
<span
  aria-checked="mixed"
  className="CheckBox TestClass"
  id="test"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="checkbox"
  tabIndex={0}
>
  Test
</span>
`;
